### DISCUSSION TITLE
Terminal token issue for ID vs Id with Wildcard

### DISCUSSION THREAD

**Discussion by brsanthu** [eclipse-langium/langium#1477] [Q&A]
We have this [DSL](https://langium.org/playground/?grammar=OYJwhgthYgBAigVwKYgJ4CgPIHYBd0EV0BZAewBNkAbALg1kdgGc8I9mBqAXgGUbkAYzy82eAFQYA3Fn7UhIsfSawA5MwHDVDJgDMAljQpduAMUPUKjABSqANKtgGjJ80YCUklbpBkIZ3wgdRjIQKhAAITRuAHkw1CiAfmC1KVVkmQw3S1gQZDxEEBxmFgJ9HGBlJgBJABFYAB9YAHVqgBlagH066SxTQKrGVR8-RzIAIwArBW4AaWQ0AHdQijjqq2swan0wZjmF5bC1incMjDjwqMG1FdRHVXG0R2dLE3mllePYWwcnC2MeO9DqsQOtPL0MEDPqCrHkCkUSqwQOVKilVLcQI4mj1MnhUBByltYHVaLAAPQAbU6YAAtAAvACCNIAWgBdKm0xksgAMNIAnKzxGSZHiQAScETWh1uvU4YViqVkRVSZTOUzmeJ2WqefzOgA6TVCmQAC30FCoOFgovFkt4uXy8sRZWVtjUjTUAB08FjPZj3aoPThHO5ODIgA&content=M4UwNiDGAuAEBCBLMZEDsDmA5AhgWxFgDMAnAez1h0kjIFc1oBuAKBdAhlgCokV0M3YuUrVaDZmw5Q4vZKkzCKVGvUat24GQnkChpZWLWSgA) to support query type of language where field can be one of these

- `BillingName` (simple field)
- `*Billing` (all fields ending with Billing)
- `*Billing*` (all fields containing Billing)
- `Billing*` (all fields starting with Billing)

We have created ID and WILD_ID terminals where WILD_ID has * char support. Out of above 4 options, first 3 are parsing fine and last one doesn't. I think it is due to lexer is consuming `Billing` part of field as ID and leaving just `*` which is not matching anything else.

How do we force lexer to treat that `Billing*` as WILD_ID instead of ID?

**Comment by brsanthu** [comment]
Updated the Regex to match WILD_ID first and then ID as in this [version](https://langium.org/playground?grammar=OYJwhgthYgBAigVwKYgJ4CgPIHYBd0EV0BZAewBNkAbALg1kdgGc8I9mBqAXgGUbkAYzy82eAFQYA3Fn7UhIsfSawA5MwHDVDJgDMAljQpduAMUPUKjABSqANKtgGjJ80YCUklbpBkIZ3wgdRjIQKhAAITRuAHkw1CiAfmC1KVVkmQw3S1gQZDxEEBxmFgJ9HGBlJgBJABFYAB9YAHVqgBlagH066SxTQKrGVR8-RzIAIwArBW4AaWQ0AHdQijjqq2swan0wZjmF5bC1incMjDjwqMG1FdRHVXG0R2dLE3mllePYWwcnC2MeO9DqsQOtPL0MEDPqCrHkCkUSqwQOVKilVLcQI4mj1MnhUBBylsWu0unVcvlCsVSsiKrRYAB6azWADanTAAFoAF4AQXZAC0AAzsgCc4gAuu5OAAdcQstlc3mCkXiyXuBrWGVyjk8-lC0USzhqpms7WKvUq6XiNUy%2BkyPEgAk4Il1On0k0K-li906pXCsXiW0YAAW%2BgoVBwsHtjqJzV45PhVKRKNotjUjTUUrwWIzmPTqilOEckpkQA&content=M4UwNiDGAuAEBCBLMZEDsDmA5AhgWxFgDMAnAez1h0kjIFc1oBuAKBdAhlgCokV0M3YuUrVaDZmw5Q4vZKkzCKVGvUat24GQnkChpZWLWSgA) and it is parsing the queries fine. 

Do you see any isses in this format?

**Reply by spoenemann** [reply]
I would use a data type rule instead of a terminal rule:
```
Field returns string:
    '*'? ID '*'?;
```
The advantage is that the asterisk `'*'` and the ID tokens are treated separately in the lexing phase, so you don't get any problems with token precedence. The default Chevrotain lexer consumes the first matching token in the list.
